client/src/document/index.tsx

```diff
@@ -366,3 +382,83 @@ function LoadingError({ error }) {
     </div>
   );
 }
+
+interface FlatFlaw {
+  name: string;
+  flaws: string[];
+  count: number;
+}
+
+function ToggleDocmentFlaws({ doc }: { doc: Doc }) {
+  const { flaws }: { flaws: object } = doc;
+  const [show, toggle] = useReducer((v) => !v, false);
+
+  useEffect(() => {
+    if (window.location.hash && window.location.hash === "#show-flaws") {
+      toggle();
+    }
+  }, []);
+
+  useEffect(() => {
+    if (show) {
+      if (!(window.location.hash && window.location.hash === "#show-flaws")) {
+        window.location.hash = "#show-flaws";
+      }
+    } else {
+      if (window.location.hash && window.location.hash === "#show-flaws") {
+        window.location.hash = "";
+      }
+    }
+  }, [show]);
+
+  let flatFlaws: FlatFlaw[] = [];
+  for (const [name, actualFlaws] of Object.entries(flaws)) {
+    flatFlaws.push({
+      name,
+      flaws: actualFlaws,
+      count: actualFlaws.length,
+    });
+  }
+  flatFlaws.sort((a, b) => b.count - a.count);
+
+  function summarizeFlaws() {
+    // Return a one-liner about all the flaws
+    const bits = flatFlaws.map((flaw) => {
+      return `${humanizeFlawName(flaw.name)}: ${flaw.count}`;
+    });
+    return bits.join(", ");
+  }
+
+  function summarizeFlawsCounts() {
+    // Return a one-liner about all the flaws
+    const bits = flatFlaws.map((flaw) => flaw.count);
+    return bits.join(" + ");
+  }
+
+  const hasFlaws = flatFlaws.length > 0;
+
+  return (
+    <div id="show-flaws" className="toggle-flaws">
+      {hasFlaws ? (
+        <button type="submit" onClick={toggle}>
+          {show ? "Hide flaws" : `Show flaws (${summarizeFlawsCounts()})`}
+        </button>
+      ) : (
+        <p className="no-flaws">
+          No known flaws at the moment
+          <span role="img" aria-label="yay!">
+            🍾
+          </span>
+        </p>
+      )}
+
+      {show ? (
+        <Suspense fallback={<div>Loading document flaws...</div>}>
+          <DocumentFlaws flaws={flatFlaws} />
+        </Suspense>
+      ) : (
+        <small>{summarizeFlaws()}</small>
+      )}
+    </div>
+  );
```

Comment:
```suggestion
    return (
    <div id="show-flaws" className="toggle-flaws">
      {flatFlaws.length > 0 ? (
        <button type="submit" onClick={toggle}>
          {show
            ? "Hide flaws"
            : `Show flaws (${flatFlaws.map((flaw) => flaw.count).join(" + ")})`}
        </button>
      ) : (
        <p className="no-flaws">
          No known flaws at the moment
          <span role="img" aria-label="yay!">
            🍾
          </span>
        </p>
      )}

      {show ? (
        <Suspense fallback={<div>Loading document flaws...</div>}>
          <DocumentFlaws flaws={flatFlaws} />
        </Suspense>
      ) : (
        <small>
          {/* a one-liner about all the flaws */}
          {flatFlaws
            .map((flaw) => `${humanizeFlawName(flaw.name)}: ${flaw.count}`)
            .join(" + ")}
        </small>
      )}
    </div>
  );
```

This is probably the most opinionated of the three suggestions. I prefer those to be inlined. Extraction is a net gain for me when
a) the function name gives new info on how to interpret the output
b) there's complexity in what's extracted, which is not essential to the larger context
For me those two aren't given, so I'd inline